// @adonisjs/validator/build/providers/ValidatorProvider.js
"use strict";
/*
 * @adonisjs/validator
 *
 * (c) Harminder Virk <virk@adonisjs.com>
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * Provider to register validator with the IoC container
 */
class ValidationProvider {
    constructor(container) {
        this.container = container;
    }
    register() {
        this.container.singleton('Adonis/Core/Validator', () => {
            return {
                validator: require('../src/Validator').validator,
                schema: require('../src/Schema').schema,
                rules: require('../src/Rules').rules,
            };
        });
    }
    async boot() {
        this.container.with(['Adonis/Core/Request', 'Adonis/Core/Validator'], (Request, Validator) => {
            require('../src/Bindings/Request').default(Request, Validator.validator.validate);
        });
    }
}
exports.default = ValidationProvider;
