// edge.js/build/src/Tags/index.js
"use strict";
/*
* edge
*
* (c) Harminder Virk <virk@adonisjs.com>
*
* For the full copyright and license information, please view the LICENSE
* file that was distributed with this source code.
*/
Object.defineProperty(exports, "__esModule", { value: true });
var If_1 = require("./If");
exports.if = If_1.ifTag;
var Else_1 = require("./Else");
exports.else = Else_1.elseTag;
var ElseIf_1 = require("./ElseIf");
exports.elseif = ElseIf_1.elseIfTag;
var Include_1 = require("./Include");
exports.include = Include_1.includeTag;
var IncludeIf_1 = require("./IncludeIf");
exports.includeIf = IncludeIf_1.includeIfTag;
var Each_1 = require("./Each");
exports.each = Each_1.eachTag;
var Component_1 = require("./Component");
exports.component = Component_1.componentTag;
var Slot_1 = require("./Slot");
exports.slot = Slot_1.slotTag;
var Debugger_1 = require("./Debugger");
exports.debugger = Debugger_1.debuggerTag;
var Set_1 = require("./Set");
exports.set = Set_1.setTag;
var Unless_1 = require("./Unless");
exports.unless = Unless_1.unlessTag;
var Yield_1 = require("./Yield");
exports.yield = Yield_1.yieldTag;
var Layout_1 = require("./Layout");
exports.layout = Layout_1.layoutTag;
var Section_1 = require("./Section");
exports.section = Section_1.sectionTag;
var Super_1 = require("./Super");
exports.super = Super_1.superTag;
