// @adonisjs/validator/build/src/utils.js
"use strict";
/*
 * @adonisjs/validator
 *
 * (c) Harminder Virk <virk@adonisjs.com>
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getArrayType = exports.getObjectType = exports.getLiteralType = exports.compileRule = void 0;
const Rules_1 = require("./Rules");
const validations = __importStar(require("./Validations"));
/**
 * Compiles the `Rule` object and returns `ParsedRule` object.
 */
function compileRule(type, subtype, rule) {
    const ruleValidation = validations[rule.name];
    if (!ruleValidation) {
        throw new Error(`"${rule.name}" rule is not registered. Use "validator.rule" to add the rule.`);
    }
    if (typeof ruleValidation.compile !== 'function') {
        throw new Error(`"${rule.name}" rule must implement the compile function`);
    }
    if (typeof ruleValidation.validate !== 'function') {
        throw new Error(`"${rule.name}" rule must implement the validate function`);
    }
    return ruleValidation.compile(type, subtype, rule.options);
}
exports.compileRule = compileRule;
/**
 * Dry function to define a literal type
 */
function getLiteralType(subtype, isOptional, ruleOptions, rules) {
    const hasSubTypeRule = rules.find((rule) => rule.name === subtype);
    return {
        getTree() {
            return {
                type: 'literal',
                subtype: subtype,
                rules: []
                    .concat(isOptional ? [] : [Rules_1.rules.required()])
                    .concat(hasSubTypeRule ? [] : [Rules_1.rules[subtype](ruleOptions)])
                    .concat(rules)
                    .map((rule) => compileRule('literal', subtype, rule)),
            };
        },
    };
}
exports.getLiteralType = getLiteralType;
/**
 * Dry function to define an object type
 */
function getObjectType(isOptional, children, rules) {
    const hasSubTypeRule = rules.find((rule) => rule.name === 'object');
    return {
        getTree() {
            return {
                type: 'object',
                rules: []
                    .concat(isOptional ? [] : [{ name: 'required', options: [] }])
                    .concat(hasSubTypeRule ? [] : [{ name: 'object', options: [] }])
                    .concat(rules)
                    .map((rule) => compileRule('object', 'object', rule)),
                children: children,
            };
        },
    };
}
exports.getObjectType = getObjectType;
/**
 * Dry function to define an array type
 */
function getArrayType(isOptional, each, rules) {
    const hasSubTypeRule = rules.find((rule) => rule.name === 'array');
    return {
        getTree() {
            return {
                type: 'array',
                rules: []
                    .concat(isOptional ? [] : [{ name: 'required', options: [] }])
                    .concat(hasSubTypeRule ? [] : [{ name: 'array', options: [] }])
                    .concat(rules)
                    .map((rule) => compileRule('array', 'array', rule)),
                ...(each ? { each } : {}),
            };
        },
    };
}
exports.getArrayType = getArrayType;
