// edge.js/build/src/Edge/index.js
"use strict";
/*
* edge
*
* (c) Harminder Virk <virk@adonisjs.com>
*
* For the full copyright and license information, please view the LICENSE
* file that was distributed with this source code.
*/
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
const Tags = __importStar(require("../Tags"));
const Loader_1 = require("../Loader");
const Context_1 = require("../Context");
const Compiler_1 = require("../Compiler");
const Renderer_1 = require("../Renderer");
/**
 * Exposes the API to render templates, register custom tags and globals
 */
class Edge {
    constructor(options = {}) {
        this.options = options;
        /**
         * Globals are shared with all rendered templates
         */
        this.GLOBALS = {};
        /**
         * List of registered tags. Adding new tags will only impact
         * this list
         */
        this.tags = {};
        /**
         * The loader to load templates. A loader can read and return
         * templates from anywhere. The default loader reads files
         * from the disk
         */
        this.loader = this.options.loader || new Loader_1.Loader();
        /**
         * The underlying compiler in use
         */
        this.compiler = new Compiler_1.Compiler(this.loader, this.tags, !!this.options.cache);
        Object.keys(Tags).forEach((name) => this.registerTag(Tags[name]));
    }
    /**
     * Mount named directory to use views. Later you can reference
     * the views from a named disk as follows.
     *
     * ```
     * edge.mount('admin', join(__dirname, 'admin'))
     *
     * edge.render('admin::filename')
     * ```
     */
    mount(diskName, dirPath) {
        if (!dirPath) {
            dirPath = diskName;
            diskName = 'default';
        }
        this.loader.mount(diskName, dirPath);
        return this;
    }
    /**
     * Un Mount a disk from the loader.
     *
     * ```js
     * edge.unmount('admin')
     * ```
     */
    unmount(diskName) {
        this.loader.unmount(diskName);
        return this;
    }
    /**
     * Add a new global to the edge globals. The globals are available
     * to all the templates.
     *
     * ```js
     * edge.global('username', 'virk')
     * edge.global('time', () => new Date().getTime())
     * ```
     */
    global(name, value) {
        this.GLOBALS[name] = value;
        return this;
    }
    /**
     * Add a new tag to the tags list.
     *
     * ```ts
     * edge.registerTag('svg', {
     *   block: false,
     *   seekable: true,
     *
     *   compile (parser, buffer, token) {
     *     const fileName = token.properties.jsArg.trim()
     *     buffer.writeRaw(fs.readFileSync(__dirname, 'assets', `${fileName}.svg`), 'utf-8')
     *   }
     * })
     * ```
     */
    registerTag(tag) {
        if (typeof (tag.run) === 'function') {
            tag.run(Context_1.Context);
        }
        this.tags[tag.tagName] = tag;
        return this;
    }
    /**
     * Register an in-memory template.
     *
     * ```ts
     * edge.registerTemplate('button', {
     *   template: `<button class="{{ this.type || 'primary' }}">
     *     @!yield($slots.main())
     *   </button>`,
     * })
     * ```
     *
     * Later you can use this template
     *
     * ```edge
     * @component('button', type = 'primary')
     *   Get started
     * @endcomponent
     * ```
     */
    registerTemplate(templatePath, contents) {
        this.loader.register(templatePath, contents);
        return this;
    }
    /**
     * Render a template with optional state
     *
     * ```ts
     * edge.render('welcome', { greeting: 'Hello world' })
     * ```
     */
    render(templatePath, state) {
        return this.getRenderer().render(templatePath, state);
    }
    /**
     * Returns a new instance of edge. The instance
     * can be used to define locals.
     */
    getRenderer() {
        return new Renderer_1.EdgeRenderer(this.compiler, this.GLOBALS);
    }
    /**
     * Share locals with the current view context.
     *
     * ```js
     * const view = edge.getRenderer()
     *
     * // local state for the current render
     * view.share({ foo: 'bar' })
     *
     * view.render('welcome')
     * ```
     */
    share(data) {
        return this.getRenderer().share(data);
    }
}
exports.Edge = Edge;
