// @adonisjs/assembler/build/src/Installer/index.js
"use strict";
/*
 * @adonisjs/assembler
 *
 * (c) Harminder Virk <virk@adonisjs.com>
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
*/
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Installer = void 0;
const execa_1 = __importDefault(require("execa"));
/**
 * Exposes the API to install production dependencies for a project
 * using yarn or npm
 */
class Installer {
    constructor(appRoot, client) {
        this.appRoot = appRoot;
        this.client = client;
    }
    /**
     * Install dependencies
     */
    async install() {
        const args = this.client === 'npm' ? ['ci', '--production'] : ['install', '--production'];
        await execa_1.default(this.client, args, {
            buffer: false,
            stdio: 'inherit',
            cwd: this.appRoot,
            env: {
                FORCE_COLOR: 'true',
            },
        });
    }
}
exports.Installer = Installer;
