// @adonisjs/sink/src/Files/index.js
"use strict";
/*
* @adonisjs/files
*
* (c) Harminder Virk <virk@adonisjs.com>
*
* For the full copyright and license information, please view the LICENSE
* file that was distributed with this source code.
*/
Object.defineProperty(exports, "__esModule", { value: true });
var File_1 = require("./Base/File");
exports.File = File_1.File;
var KeyValuePair_1 = require("./Base/KeyValuePair");
exports.KeyValuePair = KeyValuePair_1.KeyValuePair;
var Ini_1 = require("./Formats/Ini");
exports.IniFile = Ini_1.IniFile;
var Json_1 = require("./Formats/Json");
exports.JsonFile = Json_1.JsonFile;
var Mustache_1 = require("./Formats/Mustache");
exports.MustacheFile = Mustache_1.MustacheFile;
var NewLine_1 = require("./Formats/NewLine");
exports.NewLineFile = NewLine_1.NewLineFile;
var TemplateLiteral_1 = require("./Formats/TemplateLiteral");
exports.TemplateLiteralFile = TemplateLiteral_1.TemplateLiteralFile;
var Yaml_1 = require("./Formats/Yaml");
exports.YamlFile = Yaml_1.YamlFile;
var AdonisRc_1 = require("./Special/AdonisRc");
exports.AdonisRcFile = AdonisRc_1.AdonisRcFile;
var Env_1 = require("./Special/Env");
exports.EnvFile = Env_1.EnvFile;
var PackageJson_1 = require("./Special/PackageJson");
exports.PackageJsonFile = PackageJson_1.PackageJsonFile;
